internal/cli/internal/command/services/postgres/launch.go

```diff
@@ -0,0 +1,308 @@
+	AppName            string
+	ConsulUrl          string
+	ImageRef           string
+	InitialClusterSize int
+	Organization       *api.Organization
+	Password           string
+	Region             string
+	SnapshotID         string
+	VolumeSize         int
+	VMSize             string
+}
+
+func runLaunch(ctx context.Context) error {
+	initialClusterSize := flag.GetInt(ctx, "initial-cluster-size")
+	password := flag.GetString(ctx, "password")
+	volumeSize := flag.GetInt(ctx, "volume-size")
+	vmSize := flag.GetString(ctx, "vm-size")
+	snapshotID := flag.GetString(ctx, "snapshot-id")
+	consulUrl := flag.GetString(ctx, "consul-url")
+
+	name := flag.GetString(ctx, "name")
+	if name == "" {
+		if err := prompt.String(ctx, &name, "App name", "", true); err != nil {
+			return err
+		}
+	}
+
+	org, err := prompt.Org(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	region := flag.GetString(ctx, "region")
+	if region == "" {
+		var r *api.Region
+		r, err := prompt.Region(ctx)
+		if err != nil {
+			return err
+		}
+		region = r.Code
+	}
+
+	client := client.FromContext(ctx)
+
+	imageRef, err := client.API().GetLatestImageTag(ctx, "flyio/postgres")
+	if err != nil {
+		return err
+	}
+
+	config := PostgresProvisionConfig{
+		AppName:            name,
+		ConsulUrl:          consulUrl,
+		InitialClusterSize: initialClusterSize,
+		Password:           password,
+		Region:             region,
+		VolumeSize:         volumeSize,
+		VMSize:             vmSize,
+		SnapshotID:         snapshotID,
+		ImageRef:           imageRef,
+		Organization:       org,
+	}
+
+	pg := &Launch{
+		config: config,
+		client: client,
+	}
+
+	if err = pg.Launch(ctx); err != nil {
+		return fmt.Errorf("failed launching postgres: %w", err)
+	}
+
+	return nil
+}
+
+func (p *Launch) Launch(ctx context.Context) error {
+
+	app, err := p.createApp(ctx)
+	if err != nil {
+		return err
+	}
+
+	secrets, err := p.setSecrets(ctx)
+	if err != nil {
+		return err
+	}
+
+	io := iostreams.FromContext(ctx)
+	for i := 0; i < p.config.InitialClusterSize; i++ {
+		fmt.Fprintf(io.Out, "Provisioning %d of %d machines with image %s\n", i+1, p.config.InitialClusterSize, p.config.ImageRef)
+
+		machineConf, err := p.configurePostgres()
+		if err != nil {
+			return err
+		}
+
+		launchInput := api.LaunchMachineInput{
+			AppID:   app.ID,
+			OrgSlug: p.config.Organization.ID,
+			Region:  p.config.Region,
+			Config:  &machineConf,
+		}
+
+		machine, _, err := p.client.API().LaunchMachine(ctx, launchInput)
+		if err != nil {
+			return err
+		}
+
+		if err = waitForMachineState(ctx, p.client, p.config.AppName, machine.ID, "started"); err != nil {
+			return err
+		}
+	}
+
+	fmt.Fprintf(io.Out, "Provision complete\n\n")
+
+	connStr := fmt.Sprintf("postgres://postgres:%s@%s.internal:5432\n", secrets["OPERATOR_PASSWORD"], p.config.AppName)
+
+	fmt.Fprintf(io.Out, "Any app within the %s organization can connect to this Postgres using the following credentials:\n", p.config.Organization.Name)
```

Comment:
this is a block of many consecutive writes to the same output (`io.Out`). Ideally, what we'd want is for all the output to be written to a buffer first and then flushed to `io.Out`. This would keep output consistent in case, as an example, another goroutine tries to write to `os.Out` while we do.